File: app/src/main/java/ru/cities/game/fragment/MainFragment.java
package ru.cities.game.fragment;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.speech.RecognitionListener;
import android.speech.RecognizerIntent;
import android.speech.SpeechRecognizer;
import android.speech.tts.TextToSpeech;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import ru.cities.game.R;
import ru.cities.game.adapter.City;
import ru.cities.game.adapter.CityAdapter;
import ru.cities.game.db.DatabaseHelper;

public class MainFragment extends Fragment {
    private final boolean TEXT_FOUND = false;
    private final boolean TEXT_NOT_FOUND = true;
    private final List<String> calledNameCities = new ArrayList<>();
    private final ActivityResultLauncher<String> requestPermissionLauncher =
            registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
                if (isGranted) {
                    initSpeechRecognizer();
                    startListening();
                } else
                    showMessage(getString(R.string.request_rationale));
            });
    private int level;
    private int score = 0;
    private int booster = 1;
    private int increase = 0;
    private boolean lampOff = true;
    private SharedPreferences prefs;
    private CountDownTimer timer;
    private Context context;
    private TextInputEditText etText;
    private TextInputLayout lText;
    private TextView tvScore, tvTimer;
    private ImageView ivGetHint;
    private RecyclerView rvList;
    private City botCity;
    private CityAdapter adapter;
    private DatabaseHelper databaseHelper;
    private TextToSpeech textToSpeech;
    private SpeechRecognizer speechRecognizer;
    private Intent mSpeechRecognizerIntent;
    private AlertDialog alertDialog;

    public MainFragment() {
        super(R.layout.main_fragment);
    }

    private void initViews() {
        View view = getView();
        assert view != null;
        etText = view.findViewById(R.id.et_text_name);
        tvScore = view.findViewById(R.id.tv_score);
        tvTimer = view.findViewById(R.id.tv_timer);
        ivGetHint = view.findViewById(R.id.iv_get_hint);
        lText = view.findViewById(R.id.l_text_name);
        rvList = view.findViewById(R.id.rv_list);
        adapter = new CityAdapter(city -> showMessage(databaseHelper.getCityInfo(city)));
        rvList.setAdapter(adapter);
        rvList.setLayoutManager(new LinearLayoutManager(context));
    }

    private void initClickListeners() {
        ivGetHint.setOnClickListener(showHintUsageDialog);
        lText.setEndIconOnClickListener(sendSpeakText);
        etText.addTextChangedListener(setTextWatcher());
        etText.setOnEditorActionListener((textView, actionId, keyEvent) -> {
            if (actionId == EditorInfo.IME_ACTION_SEND) {
                lText.findViewById(com.google.android.material.R.id.text_input_end_icon).performClick();
                return true;
            }
            return false;
        });
    }

    private final View.OnClickListener sendSpeakText = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            if (!(context.checkSelfPermission(Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED))
                requestPermissionLauncher.launch(Manifest.permission.RECORD_AUDIO);
            else if (isNull(speechRecognizer)) {
                initSpeechRecognizer();
                startListening();
            } else startListening();
        }
    };

    private void startListening() {
        lText.setEndIconDrawable(R.drawable.ic_mic_on);
        speechRecognizer.startListening(mSpeechRecognizerIntent);
    }

    private void initSpeechRecognizer() {
        mSpeechRecognizerIntent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
        mSpeechRecognizerIntent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
        mSpeechRecognizerIntent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, Locale.getDefault());
        speechRecognizer = SpeechRecognizer.createSpeechRecognizer(context);
        speechRecognizer.setRecognitionListener(new RecognitionListener() {
            @Override
            public void onReadyForSpeech(Bundle params) {
            }

            @Override
            public void onBeginningOfSpeech() {
            }

            @Override
            public void onRmsChanged(float rms) {
            }

            @Override
            public void onBufferReceived(byte[] buffer) {
            }

            @Override
            public void onEndOfSpeech() {
                lText.setEndIconDrawable(R.drawable.ic_mic_off);
            }

            @Override
            public void onError(int error) {
                lText.setEndIconDrawable(R.drawable.ic_mic_off);
            }

            @Override
            public void onResults(Bundle results) {
                ArrayList<String> result = results.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
                if (!isNull(result)) {
                    String text = result.get(0);
                    if (text.length() < 26) {
                        etText.setText(text);
                        etText.setSelection(text.length());
                        checkInput();
                    } else showMessage(getString(R.string.speech_recognition_error));
                }
            }

            @Override
            public void onPartialResults(Bundle partialResults) {
            }

            @Override
            public void onEvent(int eventType, Bundle params) {
            }
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        final String key = "firstRun";
        if (prefs.getBoolean(key, true)) {
            timer.cancel();
            new MaterialAlertDialogBuilder(context)
                    .setTitle(getString(R.string.rules_title))
                    .setMessage(getString(R.string.rules_message))
                    .setPositiveButton(getString(R.string.rules_ok), (dialog, which) ->
                            new MaterialAlertDialogBuilder(context)
                                    .setTitle(getString(R.string.purpose_title))
                                    .setMessage(getString(R.string.purpose_message))
                                    .setPositiveButton(getString(R.string.purpose_ok), null)
                                    .show())
                    .show();
            prefs.edit().putBoolean(key, false).apply();
        }
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        this.context = context;
        databaseHelper = new DatabaseHelper(context);
        textToSpeech = new TextToSpeech(context, status -> {
            if (status == TextToSpeech.SUCCESS)
                textToSpeech.setLanguage(new Locale("ru"));
        });
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        prefs = context.getSharedPreferences("preferences", Context.MODE_PRIVATE);
        level = prefs.getInt("level", 2);
        initViews();
        initClickListeners();
        initTimer();
        timer.start();
    }

    @Override
    public void onPause() {
        if (score > prefs.getInt("record", 0))
            prefs.edit().putInt("record", score).apply();
        if (!isNull(speechRecognizer))
            speechRecognizer.cancel();
        if (!isNull(textToSpeech))
            textToSpeech.stop();
        super.onPause();
    }

    private TextWatcher setTextWatcher() {
        return new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                if (TextUtils.isEmpty(etText.getText())) {
                    lText.setEndIconOnClickListener(sendSpeakText);
                    lText.setEndIconDrawable(R.drawable.ic_mic_off);
                } else {
                    lText.setEndIconOnClickListener(sendWriteText);
                    lText.setEndIconDrawable(R.drawable.ic_send);
                }
            }

            @Override
            public void afterTextChanged(Editable editable) {
            }
        };
    }

    private final View.OnClickListener sendWriteText = view -> checkInput();

    private void checkInput() {
        String text = getValidText();
        if (isNull(text)) {
            showMessage(getString(R.string.enter_existing_city));
            return;
        }
        char firstChar = text.charAt(0);
        if (botCity != null) {
            char lastChar = botCity.getLastChar();
            if (firstChar != lastChar) {
                showMessage(getString(R.string.city_must_begin_with_letter, lastChar));
                return;
            }
        }
        if (canFindCity(firstChar, text))
            if (canFindCountry(text))
                if (canFindCityWithMistake(firstChar, text))
                    showMessage(getString(R.string.i_do_not_know_such_city));
    }

    @Nullable
    private String getValidText() {
        Editable content = etText.getText();
        if (isNull(content))
            return null;
        String text = content.toString().trim();
        if (text.isEmpty())
            return null;
        char first = text.charAt(0),
                last = text.charAt(text.length() - 1);
        if (first == 'ё' || first == 'ë' ||
                first == 'ь' || first == 'ы' ||
                first == '-' || last == '-')
            return null;
        return text.substring(0, 1).toUpperCase() + text.substring(1);
    }

    private boolean canFindCity(char firstChar, String text) {
        City userCity = databaseHelper.getCity(firstChar, text);
        if (isNull(userCity))
            return TEXT_NOT_FOUND;
        finishAdding(userCity);
        return TEXT_FOUND;
    }

    private boolean canFindCountry(String text) {
        String description = databaseHelper.getDescription(text);
        if (isNull(description))
            return TEXT_NOT_FOUND;
        showMessage(getString(R.string.is_state, text, description));
        return TEXT_FOUND;
    }

    private boolean canFindCityWithMistake(char firstChar, String text) {
        City userCity = databaseHelper.getCityWithMistake(firstChar, text);
        if (isNull(userCity))
            return TEXT_NOT_FOUND;
        finishAdding(userCity);
        return TEXT_FOUND;
    }

    private boolean canFindCalledCity(City userCity) {
        String cityName = userCity.getName();
        if (calledNameCities.contains(cityName)) {
            showMessage(getString(R.string.city_has_already_been, cityName));
            return TEXT_FOUND;
        }
        return TEXT_NOT_FOUND;
    }

    private void addPoints() {
        if (booster < increase) {
            booster *= 2;
            increase = 0;
        }
        score += booster;
        ++increase;
        tvScore.setText(String.valueOf(score));
    }

    private void finishAdding(City userCity) {
        if (canFindCalledCity(userCity)) {
            etText.setText("");
            showCity(userCity);
            addPoints();
            botSendCity(userCity);
            playSound(R.raw.send_msg);
        }
    }

    private void botSendCity(City userCity) {
        if (booster > level)
            botSurrender();
        int i = 0;
        do {
            botCity = databaseHelper.getRandomCity(userCity.getLastChar());
            if (++i > 10) botSurrender();
        } while (calledNameCities.contains(botCity.getName()));
        botCity.setType(!userCity.getType());
        showCity(botCity);
        if (!lampOff)
            switchLamp();
        if (userCity.getType())
            textToSpeech.speak(botCity.getName(), TextToSpeech.QUEUE_FLUSH, null, null);
        timer.start();
    }

    private void showCity(City city) {
        adapter.insertMessage(city);
        rvList.scrollToPosition(adapter.getItemCount() - 1);
        calledNameCities.add(city.getName());
    }

    private void botSurrender() {
        prefs.edit().putInt("level", level * 2).apply();
        prefs.edit().putBoolean("isVictory", true).apply();
        openResultFragment();
    }

    private final View.OnClickListener showHintUsageDialog = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            alertDialog = new MaterialAlertDialogBuilder(context)
                    .setTitle(getString(R.string.using_hint))
                    .setMessage(getString(R.string.spend_stars_on_hint, booster))
                    .setPositiveButton(getString(R.string.yes), (dialogInterface, i) -> {
                        if (score >= booster) {//checking whether the user has enough points
                            score -= booster;
                            tvScore.setText(String.valueOf(score));
                            botCity.setType(false);
                            botSendCity(botCity);
                            botSendCity(botCity);
                        } else
                            showMessage(getString(R.string.missing_stars, booster - score));
                    })
                    .setNegativeButton(getString(R.string.no), null).create();
            alertDialog.show();
        }
    };

    private void playSound(int soundId) {
        MediaPlayer.create(context, soundId).start();
    }

    private void showMessage(String msg) {
        Snackbar.make(requireView(), msg, Snackbar.LENGTH_LONG).setAnchorView(lText).show();
    }

    private boolean isNull(Object o) {
        return o == null;
    }

    private void initTimer() {
        timer = new CountDownTimer(60000, 1000) {
            public void onTick(long millisUntilFinished) {
                long seconds = millisUntilFinished / 1000;
                tvTimer.setText(getString(R.string.seconds, seconds));
                if (seconds < 30 && score >= booster && lampOff)
                    switchLamp();
            }

            public void onFinish() {
                prefs.edit().putBoolean("isVictory", false).apply();
                openResultFragment();
            }
        };
    }

    private void switchLamp() {
        if (lampOff)
            ivGetHint.setImageResource(R.drawable.ic_lamp_on);
        else
            ivGetHint.setImageResource(R.drawable.ic_lamp_off);
        lampOff = !lampOff;
    }

    @Override
    public void onDestroy() {
        timer.cancel();
        super.onDestroy();
    }

    private void openResultFragment() {
        if (!isNull(alertDialog))
            alertDialog.dismiss();
        prefs.edit().putInt("score", score).apply();
        requireActivity().getSupportFragmentManager().beginTransaction()
                .replace(R.id.container, ResultFragment.class, null)
                .commitAllowingStateLoss();
    }
}

File: app/src/main/java/ru/cities/game/fragment/MainFragment.kt
package ru.cities.game.fragment

import android.content.Context
import android.content.SharedPreferences
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.media.MediaPlayer
import android.os.Bundle
import android.os.CountDownTimer
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import androidx.appcompat.widget.TooltipCompat
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.internal.CheckableImageButton
import com.google.android.material.snackbar.Snackbar
import ru.cities.game.R
import ru.cities.game.adapter.*
import ru.cities.game.databinding.MainFragmentBinding
import ru.cities.game.db.DatabaseHelper
import ru.cities.game.util.*
import ru.cities.game.tts.Speech

class MainFragment : Fragment(R.layout.main_fragment) {

    private lateinit var prefs: SharedPreferences
    private lateinit var database: SQLiteDatabase
    private lateinit var adapter: MessageAdapter
    private lateinit var fragmentContext: Context
    private lateinit var fragmentActivity: FragmentActivity
    private lateinit var textToSpeech: Speech
    private var botCity: String = EMPTY_STRING
    private var level: Int = 1
    private var score: Int = 0
    private var booster: Int = 1
    private var increase: Int = 0
    private var _binding: MainFragmentBinding? = null
    private val binding get() : MainFragmentBinding = _binding!!
    private val calledCities: MutableList<String> = mutableListOf()

    /** Initialization context, activity, TTS, DB for the fast run */
    override fun onAttach(context: Context) {
        fragmentContext = requireContext()
        fragmentActivity = requireActivity()
        textToSpeech = Speech(fragmentContext)
        database = DatabaseHelper(fragmentContext).readableDatabase
        prefs = fragmentActivity.getSharedPreferences(PREFERENCES_FILE, Context.MODE_PRIVATE)
        super.onAttach(context)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        setStatusBar(fragmentActivity, 0, R.color.blue, false)
        _binding = MainFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    /** Initialization RV, CL and start timer */
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        level = prefs.getInt(KEY_LEVEL, 2)
        initRecyclerView()
        initClickListeners()
        timer.start()
    }

    /** Destroying the TTS */
    override fun onPause() {
        if (score > prefs.getInt(KEY_SCORE, 0))
            prefs.edit().putInt(KEY_SCORE, score).apply()
        textToSpeech.destroy()
        super.onPause()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        //database.close()
        //timer.cancel()
        //textToSpeech.destroy()
        _binding = null
    }

    /** Set custom adapter */
    private fun initRecyclerView() {
        adapter = MessageAdapter()
        binding.rvList.adapter = adapter
        binding.rvList.layoutManager = LinearLayoutManager(fragmentContext)
    }

    private fun initClickListeners() {
        binding.ivGetHint.setOnClickListener(showHintUsageDialog)
        binding.lCityName.setEndIconOnClickListener(userSendCity)
        TooltipCompat.setTooltipText(
            binding.ivGetHint,
            getString(R.string.get_hint)
        )//notification that the user may receive a hint

        binding.etCityName.setOnEditorActionListener { _, actionId, _ ->
            return@setOnEditorActionListener if (actionId == EditorInfo.IME_ACTION_SEND) {
                binding.lCityName.findViewById<CheckableImageButton>(com.google.android.material.R.id.text_input_end_icon)
                    .performClick()
                true
            } else false
        }
    }

    private var timer = object : CountDownTimer(60000, 1000) {
        override fun onTick(millisUntilFinished: Long) {
            val seconds = millisUntilFinished / 1000//converting milliseconds to
            binding.tvTimer.text = getString(R.string.seconds, seconds.toString())
            if (seconds == TIME_USE_HINT && score >= booster)
                binding.ivGetHint.performLongClick()//call notification
        }

        override fun onFinish() {
            openResultFragment(false)//at the end of the countdown, a fragment of defeat opens
        }
    }

    /** The user wants to get a hint */
    private val showHintUsageDialog = View.OnClickListener {
        MaterialAlertDialogBuilder(fragmentContext)
            .setTitle(getString(R.string.using_hint))
            .setMessage(getString(R.string.spend_stars_on_hint, booster))
            .setPositiveButton(getString(R.string.yes)) { _, _ ->
                if (score >= booster) {//checking whether the user has enough points
                    score -= booster
                    binding.tvScore.text = score.toString()
                    botSendCity(USER_ID, botCity)//the bot responds for the user
                    botSendCity(BOT_ID, botCity)//the bot responds for itself
                    timer.start()//restarting the countdown
                } else
                    showMessage(getString(R.string.missing_stars, booster - score))
            }
            .setNegativeButton(getString(R.string.no), null).show()
    }

    /** Displaying a message */
    private fun showMessage(msg: String) {
        Snackbar.make(requireView(), msg, Snackbar.LENGTH_SHORT).setAnchorView(binding.lCityName)
            .show()
    }

    /** Processing of the city entered by the user */
    private val userSendCity = View.OnClickListener {
        val userCity: String =
            normalizeUserCity(binding.etCityName.text.toString())//checking the correctness of the entered text
        val firstChar: Char = userCity.first()
        if (userCity.isBlank() || firstChar == 'Ь' || firstChar == 'Ы') {
            showMessage(getString(R.string.enter_existing_city))
            return@OnClickListener//return CITY_FOUND
        }
        if (botCity.isNotEmpty()) {//checking the city when launching the app
            val lastChar: Char = getLastChar(botCity)
            if (firstChar != lastChar) {//checking that the city should start with the last letter
                showMessage(getString(R.string.the_city_must_begin_with_letter, lastChar))
                return@OnClickListener//return CITY_FOUND
            }
        }
        if (canFoundCity(firstChar, userCity))
            if (canFoundModifiedCity(firstChar, userCity))
                if (canFoundOldCity(userCity))
                    if (canFoundCountry(userCity))
                        if (canFoundCityWithMistake(firstChar, userCity))
                            showMessage(getString(R.string.i_do_not_know_such_city))
    }

    /** Search for cities in the database with one error */
    private fun canFoundCityWithMistake(firstChar: Char, userCity: String): Boolean {
        var queryString = "SELECT * FROM $firstChar WHERE name = '$firstChar'"
        val userCityLength = userCity.length
        for (i in 1 until userCityLength) {
            val chars = userCity.toMutableList()
            chars[i] = '_'
            queryString += " OR name LIKE '${chars.joinToString(separator = "")}'"
        }
        return canFoundCity(firstChar, userCity, queryString)
    }

    /** Search cities in the database */
    private fun canFoundCity(
        firstChar: Char,
        userCity: String,
        queryString: String = "SELECT * FROM $firstChar WHERE name = '$userCity'"
    ): Boolean {
        val cursor: Cursor =
            database.rawQuery(
                queryString,
                null
            )
        if (cursor.moveToFirst()) {//checking the receipt of data
            //there was an error with resending a city containing the letter "ё"
            val city: String = cursor.getString(0)
            if (calledCities.firstOrNull { it == city } != null) {//checking whether this city wasn't there before
                showMessage(getString(R.string.city_has_already_been, city))
                return CITY_FOUND
            }
            binding.etCityName.setText(EMPTY_STRING)
            timer.start()
            val flag: Int = cursor.getInt(1)
            cursor.close()
            playSoundMessage()
            addCity(USER_ID, city, flag)
            addPoints()
            botSendCity(BOT_ID, city)
            return CITY_FOUND
        }
        cursor.close()
        return CITY_NOT_FOUND
    }

    private fun addPoints() {
        if (booster < increase) {
            booster = booster++ * 2
            increase = 0//the counter for increasing the booster
        }
        score += booster
        increase++
        binding.tvScore.text = score.toString()
    }

    /** Search for the only letter 'ё' in the word */
    private fun canFoundModifiedCity(firstChar: Char, userCity: String): Boolean {
        for (index in userCity.indices)
            if (userCity[index] == 'е')
                if (!canFoundCity(//or canFoundCityWithMistake?
                        firstChar,
                        userCity.substring(0, index) +
                                'ё' + userCity.substring(index + 1)
                    )
                )
                    return CITY_FOUND
        return CITY_NOT_FOUND
    }

    /** Playback of the message sound */
    private fun playSoundMessage() {
        MediaPlayer.create(fragmentActivity, R.raw.send_msg).start()
    }

    /** Search for a city in the table of old city names */
    private fun canFoundOldCity(userCity: String): Boolean {
        val cursor: Cursor =
            database.rawQuery("SELECT * FROM old_city WHERE old_name = '$userCity'", null)
        if (cursor.moveToFirst()) {
            val oldCity: String = cursor.getString(0)
            showMessage(getString(R.string.former_name, userCity, oldCity))
            cursor.close()
            return CITY_FOUND
        }
        cursor.close()
        return CITY_NOT_FOUND
    }

    /** Search for a city in the country table */
    private fun canFoundCountry(userCity: String): Boolean {
        val cursor: Cursor =
            database.rawQuery("SELECT * FROM country WHERE name = '$userCity'", null)
        if (cursor.moveToFirst()) {
            showMessage(getString(R.string.is_country, userCity))
            cursor.close()
            return CITY_FOUND
        }
        cursor.close()
        return CITY_NOT_FOUND
    }

    /** Adding a city to the RV and to the list of called cities */
    private fun addCity(id: Boolean, city: String, flag: Int) {
        val cityModel = MessageModel(id, city, flag)
        adapter.insertMessage(cityModel)
        binding.rvList.scrollToPosition(adapter.itemCount - 1)//scroll down the RV
        calledCities.add(city)
    }

    /** The bot responds to the user */
    private fun botSendCity(id: Boolean, userCity: String) {
        if (booster > level)
            botSurrender()
        var flag: Int
        var numberSamples = 0
        var cursor: Cursor
        do {
            if (numberSamples++ > MAX_NUMBER_SAMPLES)//checking for the maximum number of repeated samples
                botSurrender()
            cursor =
                database.rawQuery(
                    "SELECT * FROM ${getLastChar(userCity)} ORDER BY RANDOM() LIMIT 1",
                    null
                )
            cursor.moveToFirst()
            botCity = cursor.getString(0)
            flag = cursor.getInt(1)
        } while (calledCities.firstOrNull { it == botCity } != null)//checking whether there has already been such a city
        cursor.close()
        if (id == BOT_ID)//checking that the bot speaks for itself
            textToSpeech = Speech(fragmentContext, botCity)
        addCity(id, botCity, flag)
    }

    /** The bot capitulates */
    private fun botSurrender() {
        prefs.edit().putInt(KEY_LEVEL, level * 2).apply()
        openResultFragment(true)
    }

    /** Opening the fragment of victory or defeat */
    fun openResultFragment(isVictory: Boolean) {
        timer.cancel()
        val fragment = ResultFragment()
        fragment.arguments = bundleOf(
            SCORE_ID to score,
            IS_VICTORY_ID to isVictory
        )
        fragmentActivity.supportFragmentManager.beginTransaction()
            .replace(R.id.container, fragment).commit()
    }
}
